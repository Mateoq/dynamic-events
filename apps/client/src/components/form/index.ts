export * from './Button';
export * from './SelectInput';
export * from './TextInput';
export * from './DatePicker';
export * from './CheckBox';
