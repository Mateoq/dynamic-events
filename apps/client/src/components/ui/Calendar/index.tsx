import React from 'react';

import { CalendarMonth } from './CalendarMonth';
import { CalendarWeek } from './CalendarWeek';
import { CalendarDay } from './CalendarDay';
import { CalendarNav } from './CalendarNav';
import { useCalendarStateStore } from '@/stores/calendar-state-store';
import { CalendarData, Direction, CalendarType } from '@/types';
import {
  getYearKey,
  getMonthKey,
  getWeekKey,
  getDayKey,
  getMonthFromState,
  getWeekFromState,
  getDayFromState
} from '@/utils/calendar';

export type CalendarProps = {
  type: CalendarType;
  data: CalendarData;
  onNav: (direction: Direction) => void;
  onCreateEvent: () => void;
  onChangeView: (type: CalendarType) => void;
  onClickEvent: (id: string) => void;
};

export const Calendar: React.FC<CalendarProps> = (props) => {
  const { type, data, onNav, onCreateEvent, onChangeView, onClickEvent } =
    props;
  const state = useCalendarStateStore((state) => state.calendarState);
  let currentComponent = (
    <div className="text-red-600 text-xl">Unknown Calendar</div>
  );

  console.log('DATA', data);

  if (state.year === 0 || state.month === 0 || state.day === 0) {
    return null;
  }

  switch (type) {
    case CalendarType.MOTH:
      const monthData = getMonthFromState(data, state);
      currentComponent = (
        <CalendarMonth data={monthData} onClickEvent={onClickEvent} />
      );
      break;
    case CalendarType.WEEK:
      const weekData = getWeekFromState(data, state);
      currentComponent = (
        <CalendarWeek data={weekData} onClickEvent={onClickEvent} />
      );
      break;
    case CalendarType.DAY:
      const dayData = getDayFromState(data, state);
      currentComponent = (
        <CalendarDay data={dayData} onClickEvent={onClickEvent} />
      );
      break;
  }

  return (
    <div className="w-full">
      <div className="w-full flex flex-col">
        <CalendarNav
          currentView={type}
          onNav={onNav}
          onCreateEvent={onCreateEvent}
          onChangeView={onChangeView}
        />
        {currentComponent}
      </div>
    </div>
  );
};
