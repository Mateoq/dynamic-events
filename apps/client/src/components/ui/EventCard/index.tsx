import React from 'react';

import { Event } from '@/types';

export type EventCardProps = {
  data: Event;
  onClick?: (id: string) => void;
};

export const EventCard: React.FC<EventCardProps> = (props) => {
  const { data, onClick } = props;
  return (
    <button
      className="flex-1 px-1 overflow-hidden"
      style={{ background: data.color }}
      onClick={() => onClick?.(data.id)}
    >
      <p className="text-white text-sm truncate">{data.title}</p>
    </button>
  );
};
