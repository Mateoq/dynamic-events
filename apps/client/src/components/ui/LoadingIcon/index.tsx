import React from 'react';
import { motion } from 'framer-motion';
import { FaSpinner } from 'react-icons/fa6';

export type LoadingIconProps = {
  color?: string;
};

export const LoadingIcon: React.FC<LoadingIconProps> = (props) => {
  const { color = 'white' } = props;
  return (
    <motion.div
      className="w-4 h-4 round-full"
      animate={{ rotate: [0, 360] }}
      transition={{ duration: 1, ease: 'linear', repeat: Infinity }}
    >
      <FaSpinner color={color} />
    </motion.div>
  );
};
