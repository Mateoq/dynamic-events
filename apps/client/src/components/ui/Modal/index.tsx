import React from 'react';
import { AnimatePresence, motion } from 'framer-motion';
import { FaXmark } from 'react-icons/fa6';

export type ModalProps = {
  isOpen?: boolean;
  title: string;
  onClose: () => void;
};

export const Modal: React.FC<React.PropsWithChildren<ModalProps>> = (props) => {
  const { isOpen, title, children, onClose } = props;

  return (
    <AnimatePresence>
      {isOpen && (
        <motion.div
          initial={{ opacity: 0 }}
          animate={{ opacity: 1 }}
          exit={{ opacity: 0 }}
          className="w-screen h-screen fixed top-0 left-0 flex items-center justify-center z-50 overflow-visible"
        >
          <div className="w-[500px] h-fit bg-white rounded shadow-2xl relative flex flex-col">
            <div className="flex justify-between h-fit py-2 px-2 bg-gray-500 rounded-t">
              <h3 className="text-lg text-gray-50 font-medium">{title}</h3>
              <button
                onClick={onClose}
                aria-label="Close Modal"
                className="h-6 w-6 transition-opacity hover:opacity-80 active:opacity-100"
              >
                <FaXmark size="100%" className="text-gray-50" />
              </button>
            </div>
            <div className="px-3 py-5">{children}</div>
          </div>
        </motion.div>
      )}
    </AnimatePresence>
  );
};
