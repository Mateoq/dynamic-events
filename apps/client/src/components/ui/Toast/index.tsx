import React from 'react';
import { AnimatePresence, motion } from 'framer-motion';
import clsx from 'clsx';

import { ToastType } from '@/types';

export type ToastProps = {
  isOpen?: boolean;
  type: ToastType;
  text: string;
};

export const Toast: React.FC<ToastProps> = (props) => {
  const { isOpen, type, text } = props;
  const isSuccess = type === ToastType.SUCCESS;
  const isError = type === ToastType.ERROR;

  return (
    <AnimatePresence>
      {isOpen && (
        <motion.div
          initial={{ x: '100%', opacity: 0 }}
          animate={{ x: 0, opacity: 1 }}
          exit={{ x: '100%', opacity: 0 }}
          transition={{ duration: 0.2 }}
          className={clsx(
            'w-52 min-h-24 rounded filter drop-shadow-lg flex items-center justify-center p-5 z-[100] fixed bottom-5 right-5',
            { 'bg-gray-900': isSuccess, 'bg-red-600': isError }
          )}
        >
          <span className="text-md text-gray-100">{text}</span>
        </motion.div>
      )}
    </AnimatePresence>
  );
};
