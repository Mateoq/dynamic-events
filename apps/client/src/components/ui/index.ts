export * from './Calendar';
export * from './LoadingIcon';
export * from './Modal';
export * from './Toast';
export * from './EventCard';
