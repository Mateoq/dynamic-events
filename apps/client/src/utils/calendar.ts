import {
  Time,
  Day,
  Month,
  Week,
  Year,
  CalendarData,
  CalendarState,
  Direction
} from '@/types';
import {
  MAX_CALENDAR_MONTH_INDEX,
  MAX_CALENDAR_MONTH,
  MAX_CALENDAR_WEEKS,
  MAX_CALENDAR_WEEK_DAYS,
  MIN_DAY_HOUR,
  MAX_DAY_HOUR,
  HOUR_MINUTES_STEP,
  MINUTE_SECTIONS,
  WEEKDAYS
} from '@/constants';

export function checkIsToday(date: Date) {
  const todayDate = new Date();
  return (
    todayDate.getFullYear() === date.getFullYear() &&
    todayDate.getMonth() === date.getMonth() &&
    todayDate.getDate() === date.getDate()
  );
}

export function getYearKey(year: number): string {
  return `${year}`;
}

export function getMonthKey(year: number, month: number): string {
  return `${year}-${month}`;
}

export function getWeekKey(year: number, month: number, week: number): string {
  return `${year}-${month}-${week}`;
}

export function getDayKey(
  year: number,
  month: number,
  week: number,
  day: number
): string {
  return `${year}-${month}-${week}-${day}`;
}

export function getFirstMonthDate(year: number, month: number): number {
  return new Date(year, month, 1).getDay();
}

export function getLastMonthDate(year: number, month: number): number {
  return new Date(year, month + 1, 0).getDate();
}

export function getLastMonthWeekDate(year: number, month: number): number {
  return new Date(year, month + 1, 0).getDay();
}

export function createTimeData(hour: number): Time {
  const key = `${hour}`;
  const minutes: string[] = [...Array(MINUTE_SECTIONS)].map(
    (_, index) => `${hour}:${index * HOUR_MINUTES_STEP}`
  );

  return {
    key,
    hour,
    minutes
  };
}

export function createDayData(
  year: number,
  month: number,
  week: number,
  day: number
): Day {
  const key = getDayKey(year, month, week, day);
  const date = new Date(year, month, day);
  const isToday = checkIsToday(date);
  const times: Record<string, Time> = {};

  for (let hour = MIN_DAY_HOUR; hour < MAX_DAY_HOUR; hour++) {
    times[`${hour}`] = createTimeData(hour);
  }

  return {
    key,
    value: day,
    isToday,
    times
  };
}

export function createWeekData(
  year: number,
  month: number,
  week: number
): Week {
  const key = getWeekKey(year, month, week);
  const monthFirstIndex = getFirstMonthDate(year, month);
  const firstIndex = week === 0 ? monthFirstIndex : 0;
  const lastIndex =
    week === MAX_CALENDAR_WEEKS - 1
      ? getLastMonthWeekDate(year, month)
      : MAX_CALENDAR_WEEK_DAYS - 1;
  const days: Record<string, Day> = {};
  const daysOffset = firstIndex;

  for (let dayI = 0; dayI < MAX_CALENDAR_WEEK_DAYS; dayI++) {
    let day = 0;
    if (week === 0) {
      if (dayI > MAX_CALENDAR_WEEK_DAYS - firstIndex - 1) break;
      day = dayI + 1;
    } else {
      if (week === MAX_CALENDAR_WEEKS - 1 && dayI > lastIndex) break;
      day = week * MAX_CALENDAR_WEEK_DAYS - (monthFirstIndex - 1) + dayI;
    }

    days[getDayKey(year, month, week, day)] = createDayData(
      year,
      month,
      week,
      day
    );
  }

  return {
    key,
    value: week,
    firstIndex,
    lastIndex,
    days
  };
}

export function createMonthData(year: number, month: number): Month {
  const firstIndex = getFirstMonthDate(year, month);
  const lastIndex = getLastMonthDate(year, month);
  const key = getMonthKey(year, month);
  const weeks: Record<string, Week> = {};

  for (let week = 0; week < MAX_CALENDAR_WEEKS; week++) {
    weeks[getWeekKey(year, month, week)] = createWeekData(year, month, week);
  }

  return {
    key,
    value: month,
    firstIndex,
    lastIndex,
    weeks
  };
}

export function createYearData(year: number): Year {
  const key = getYearKey(year);
  const months: Record<string, Month> = {};

  for (let month = 0; month < MAX_CALENDAR_MONTH; month++) {
    months[getMonthKey(year, month)] = createMonthData(year, month);
  }

  return {
    key,
    value: year,
    months
  };
}

export function initCalendarData(initialDate = new Date()): CalendarData {
  const initialYear = initialDate.getFullYear();
  const yearData = createYearData(initialYear);
  const yearMap: Record<string, Year> = {
    [getYearKey(initialYear)]: yearData
  };

  return {
    years: yearMap
  };

  // const key = `${initialDate.getFullYear()}-${initialDate.getMonth()}`;
  // const monthData = mapMonthData(initialDate);
  // const calendarData = {
  //   [key]: monthData
  // };
  // return {
  //   currentKey: key,
  //   months: calendarData,
  //   currentDate: {
  //     year: initialDate.getFullYear(),
  //     month: initialDate.getMonth()
  //   }
  // };
}

export function mapMonthDays(data: Month): Array<null | Day> {
  const days: Day[] = [];
  const result: Array<null | Day> = [];

  for (const week of Object.values(data.weeks)) {
    days.push(...Object.values(week.days));
  }

  let dayIndex = 0;
  for (let i = 0; i < MAX_CALENDAR_MONTH_INDEX; i++) {
    if (i < data.firstIndex || i > days.length - 1) {
      result.push(null);
      continue;
    }

    result.push(days[dayIndex]);
    dayIndex++;
  }

  return result;
}

export function mapWeekDays(data: Week): Array<null | Day> {
  const days = Object.values(data.days);
  const result: Array<null | Day> = [];

  let dayIndex = 0;
  for (let i = 0; i < MAX_CALENDAR_WEEK_DAYS; i++) {
    if (i < data.firstIndex || dayIndex > days.length - 1) {
      result.push(null);
      continue;
    }

    result.push(days[dayIndex]);
    dayIndex++;
  }

  return result;
}

export function findDayWeek(calendarData: CalendarData, date: Date): number {
  const year = date.getFullYear();
  const month = date.getMonth();
  const day = date.getDate();
  const weekResult = Object.values(
    calendarData.years[getYearKey(year)].months[getMonthKey(year, month)].weeks
  ).find((week) => Object.values(week.days).some((d: Day) => d.value === day))!;

  return weekResult.value;
}

export function getWeekDay(year: number, month: number, day: number): string {
  const date = new Date(year, month, day);
  const weekDay = WEEKDAYS[date.getDay()];

  return weekDay;
}

export function getMonthFromState(data: CalendarData, state: CalendarState) {
  const { year, month } = state;
  return data.years[getYearKey(year)].months[getMonthKey(year, month)];
}

export function getWeekFromState(
  data: CalendarData,
  state: CalendarState
): Week {
  const { year, month, week } = state;
  return data.years[getYearKey(year)].months[getMonthKey(year, month)].weeks[
    getWeekKey(year, month, week)
  ];
}

export function getDayFromState(data: CalendarData, state: CalendarState): Day {
  const { year, month, week, day } = state;
  return data.years[getYearKey(year)].months[getMonthKey(year, month)].weeks[
    getWeekKey(year, month, week)
  ].days[getDayKey(year, month, week, day)];
}

export function computeMonthState(
  direction: Direction,
  state: CalendarState,
  data: CalendarData
) {
  const newState = { ...state };
  newState.month =
    direction === Direction.LEFT ? newState.month - 1 : newState.month + 1;

  if (newState.month >= MAX_CALENDAR_MONTH) {
    newState.year += 1;
    newState.month = 0;
    newState.week = 0;
    newState.day = 1;
  } else if (newState.month < 0) {
    newState.year -= 1;
    newState.month = MAX_CALENDAR_MONTH - 1;
    newState.week = MAX_CALENDAR_WEEKS - 1;
    newState.day = Object.values(
      getWeekFromState(data, newState).days
    )[0].value;
  }

  return newState;
}

export function propagateStateChangeWeek(
  state: CalendarState,
  data: CalendarData
) {
  const newState = { ...state };
  if (newState.week >= MAX_CALENDAR_WEEKS) {
    newState.week = 0;
    newState.day = 1;

    if (newState.month + 1 >= MAX_CALENDAR_MONTH) {
      newState.month = 0;
      newState.year += 1;
    } else {
      newState.month += 1;
    }
  } else if (newState.week < 0) {
    newState.week = MAX_CALENDAR_WEEKS - 1;

    if (newState.month === 0) {
      newState.month = MAX_CALENDAR_MONTH - 1;
      newState.year -= 1;
    } else {
      newState.month -= 1;
    }

    const weekDays = Object.values(getWeekFromState(data, newState).days);
    newState.day = weekDays[weekDays.length - 1].value;
  }

  return newState;
}

export function computeWeekState(
  direction: Direction,
  state: CalendarState,
  data: CalendarData
) {
  let newState = { ...state };
  newState.week =
    direction === Direction.LEFT ? newState.week - 1 : newState.week + 1;

  newState = propagateStateChangeWeek(newState, data);

  if (newState.day !== 1) {
    const weekDays = Object.values(getWeekFromState(data, newState).days);
    newState.day = weekDays[weekDays.length - 1].value;
  }

  return newState;
}

export function computeDayState(
  direction: Direction,
  state: CalendarState,
  data: CalendarData
) {
  let newState = { ...state };
  const currentMonth = getMonthFromState(data, newState);
  newState.day =
    direction === Direction.LEFT ? newState.day - 1 : newState.day + 1;

  if (newState.day > currentMonth.lastIndex) {
    newState.day = 1;
    newState.week = 0;

    if (newState.month + 1 >= MAX_CALENDAR_MONTH) {
      newState.month = 0;
      newState.year += 1;
    } else {
      newState.month += 1;
    }
  } else if (newState.day < 1) {
    if (newState.month === 0) {
      newState.month = MAX_CALENDAR_MONTH - 1;
      newState.year -= 1;
    } else {
      newState.month -= 1;
    }
    const newMonth = getMonthFromState(data, state);
    newState.week = MAX_CALENDAR_WEEKS - 1;
    newState.day = newMonth.lastIndex;
  }

  const weekDays = Object.values(getWeekFromState(data, newState).days);
  if (newState.day < weekDays[0].value) {
    newState.week -= 1;
    newState = propagateStateChangeWeek(newState, data);
  } else if (newState.day > weekDays[weekDays.length - 1].value) {
    newState.week += 1;
    newState = propagateStateChangeWeek(newState, data);
  }

  return newState;
}
