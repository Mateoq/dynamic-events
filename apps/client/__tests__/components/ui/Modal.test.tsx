import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';

import { Modal } from '@/components/ui';
import { sleep } from '@/utils/functions';

describe('Modal', () => {
  it('should match snapshot', async () => {
    const { container } = render(
      <Modal isOpen title="Title" onClose={() => {}} />
    );

    // Wait for the fade in animation to finish
    await sleep(1000);
    const title = await screen.findByText('Title');

    expect(title).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
