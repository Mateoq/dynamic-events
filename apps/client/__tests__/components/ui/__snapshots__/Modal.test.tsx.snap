// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should match snapshot 1`] = `
<div>
  <div
    class="w-screen h-screen fixed top-0 left-0 flex items-center justify-center z-50 overflow-visible"
    style="opacity: 1; will-change: transform;"
  >
    <div
      class="w-[500px] h-fit bg-white rounded shadow-2xl relative flex flex-col"
    >
      <div
        class="flex justify-between h-fit py-2 px-2 bg-gray-500 rounded-t"
      >
        <h3
          class="text-lg text-gray-50 font-medium"
        >
          Title
        </h3>
        <button
          aria-label="Close Modal"
          class="h-6 w-6 transition-opacity hover:opacity-80 active:opacity-100"
        >
          <svg
            class="text-gray-50"
            fill="currentColor"
            height="100%"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 384 512"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
            />
          </svg>
        </button>
      </div>
      <div
        class="px-3 py-5"
      />
    </div>
  </div>
</div>
`;
